### ISSUE TITLE
[SOURCE] Slack: New Event Sources

### ISSUE THREAD

**Issue by michelle0927** [PipedreamHQ/pipedream#1010]
**Describe the event source. What app, what event?**
Slack

**New Public Message Anywhere** | Triggers when a new message is posted to any public channel.
**New Mention** | Triggers when a username or highlight word is mentioned in a public channel.
**New Saved Message** | Triggers when you save a message.
**New Channel** | Triggers when a new channel is created.
**New User** | Triggers when a new Slack user is created / first joins your organization.
**New Message to Private Channel** | Triggers when a new message is posted to a private channel.
**New Team Custom Emoji** | Triggers when a custom emoji has been added to a team.
**New Reaction Added** | Triggers when a reaction is added to a message in a public channel.
**New File** | Triggers when a new file is uploaded to your workspace.
**New Direct Message** | Triggers when a new message is added to a direct message channel.
**New Multiparty Direct Message** | Triggers when a new message is added to a multiparty direct message channel.
**New Event** | Triggers when a new message or other events are created.

**Please provide a link to the relevant API docs for the specific service / operation this trigger is tied to.**
https://api.slack.com/reference

Please work on #862 with this, as well.

**Comment by masonwoodford** [comment]
I will try working on this as well.

**Comment by dylburger** [comment]
Huge thanks for all the help @masonwoodford !

**Comment by masonwoodford** [comment]
I'm really having trouble wrapping my mind around how these event sources work, I think I might be understanding the idea behind them wrong. To me it seemed like you would use the Slack event API to subscribe to an event source, but in the current Slack event source the code just lets the user select a conversation. I don't see where the run() function actually gets called with the event in any way. What I'm thinking is that somehow the code maybe listens to all events emitted by that particular conversation and then acts only the relevant ones, but I'm really not sure. If you have any free time to give a short explanation on where I'm going wrong with my thought process I would very much appreciate it, thank you!

**Comment by dylburger** [comment]
@masonwoodford yes I apologize, I should have been more clear about that since Slack is a unique model.

Many apps, like GitHub or Google, support user-specific subscription webhooks. So when a user creates a source, you can make an API call to create a webhook subscription using their credentials, and point it to the HTTP endpoint for that source. This way only the user's data will be directed to that specific source. See [here](https://github.com/PipedreamHQ/pipedream/blob/130a5264245230e39a8aaee6879a9a5de4ece76f/components/github/github.app.js#L351) and [here](https://github.com/PipedreamHQ/pipedream/blob/130a5264245230e39a8aaee6879a9a5de4ece76f/components/google_drive/google_drive.app.js#L384) for examples.

Other apps don't support user-specific webhook subscriptions, and direct all webhooks to a single HTTP endpoint owned by the app developer. This is the case for Slack. So Pipedream operates a service that ingests all webhooks, for all users, and directs them to the right, user-specific source. This code in the `new-messages-in-channel` source:

```javascript
slackApphook: {
  type: "$.interface.apphook",
  appProp: "slack",
  async eventNames() {
    return this.conversations || []
  },
},
```

is responsible for listening for events emitted by that internal service. This tells the source to listen for events specific to the authenticated user (implicit), and for events that match the values returned by the `eventNames` function. In this case, our internal service emits events tied to the name of the channel. When a user creates this source, they have to select channels they want to watch, and the source will only run when a new message arrives for that channel.

Much of the logic for processing, filtering, and routing events to the right user is contained in our internal service, so the run method of [the `new-messages-in-channel` source](https://github.com/PipedreamHQ/pipedream/blob/130a5264245230e39a8aaee6879a9a5de4ece76f/components/slack/sources/new-message-in-channels/new-message-in-channels.js#L119) will run for every new message in the target channels. The source just has to format the event and emit it with the proper `id`, `summary`, and other metadata.

I'm adding the **blocked** label to this issue, since you won't be able to work on these sources without us adding logic to our internal service to emit these events, first. I'm looking into that now and I'll let you know as we make progress here.

**Comment by masonwoodford** [comment]
Ok, sounds good, thanks for the in depth explanation!

**Comment by js07** [comment]
I created a separate issue, https://github.com/PipedreamHQ/pipedream/issues/3148, to support a subset of these sources which should be unblocked.

**Comment by nebbles** [comment]
> New Saved Message | Triggers when you save a message.

Hit up against the lack of this one today when trying to build a workflow. Unfortunately, it seems like a regression since #3148 due to Slack moving away from 'stars' and to their new 'Save' feature.

**Comment by malexanderlim** [comment]
@nebbles this is something that we ran up against ourselves; Slack deprecated functionality for their Stars API methods and endpoints without any replacement, which is unfortunate.

https://api.slack.com/changelog/2023-07-its-later-already-for-stars-and-reminders#how
> If your app relies on these methods, we recommend retiring the functionality in your app. APIs for Save it for Later are not yet available and we don’t have a timeline for when or if they'll be made available.

We'll be monitoring and if/when API endpoints for Save it For Later become available, we'll be sure to get those integrated ASAP.

**Comment by nebbles** [comment]
@malexanderlim thanks for following up. I remember this botched release, as it also affected a productivity app I use, Akiflow. 

The bizarre thing is that Akiflow managed to get this addressed within a month or two of the break — and it seems Zapier may have done too. See https://zapier.com/apps/email/integrations/slack/1590/send-yourself-an-email-for-new-saved-slack-messages (which is the exact use case I'm trying to build in Pipedream)

(Happy to open a new issue for this if it helps)

**Comment by malexanderlim** [comment]
> @malexanderlim thanks for following up. I remember this botched release, as it also affected a productivity app I use, Akiflow.
> 
> The bizarre thing is that Akiflow managed to get this addressed within a month or two of the break — and it seems Zapier may have done too. See https://zapier.com/apps/email/integrations/slack/1590/send-yourself-an-email-for-new-saved-slack-messages (which is the exact use case I'm trying to build in Pipedream)
> 
> (Happy to open a new issue for this if it helps)

@nebbles thanks for sharing this additional context - if you don't mind creating a new issue to track this, I can spend some time researching exactly how they might have been able to accomplish this - would definitely like to carry-over the functionality if there is some way to do this!

**Comment by malexanderlim** [comment]
@nebbles tracking here: https://github.com/PipedreamHQ/pipedream/issues/12436

**Comment by malexanderlim** [comment]
@nebbles , the New Saved Message source has been landed here: https://github.com/PipedreamHQ/pipedream/issues/12436

Please give it a try and let us know what you think!
